vochain/state/sik.go

```diff
@@ -0,0 +1,128 @@
+// hysteresis has
+const hysteresisLen = 32
+
+// SetSIK function creates or update the SIK of the provided address in the
+// state. It covers the following cases:
+//   - It checks if already exists a valid SIK for the provided address and if
+//     so it returns an error.
+//   - If no SIK exists for the provided address, it will create one with the
+//     value provided.
+//   - If exists a logically deleted SIK for the address provided, checks the
+//     hysteresis height for that address before update its SIK.
+//   - If the hysteresis height is not reached yet, it returns an error.
+//   - If the hysteresis height is reached, it updates the value of the sik with
+//     the provided one.
+func (v *State) SetSIK(address common.Address, newSik []byte) error {
+	// check if exists a registered sik for the provided address, query also for
+	// no commited tree version
+	sik, err := v.mainTreeViewer(false).DeepGet(address.Bytes(), StateTreeCfg(TreeSIK))
+	if errors.Is(err, arbo.ErrKeyNotFound) {
+		// if not exists create it
+		log.Debugw("setSIK (create)",
+			"address", address.String(),
+			"sik", hex.EncodeToString(sik))
+		v.Tx.Lock()
+		defer v.Tx.Unlock()
+		return v.Tx.DeepAdd(address.Bytes(), newSik, StateTreeCfg(TreeSIK))
+	} else if err != nil {
+		return err
+	}
+	// check if is a valid sik
+	if validSIK(sik) {
+		return ErrRegisteredValidSIK
+	}
+	// check if the hysteresis height is reached
+	hysteresis := decodeHysteresis(sik)
+	if hysteresis > v.CurrentHeight() {
+		return ErrHysteresisNotReached
+	}
+	log.Debugw("setSIK (update)",
+		"address", address.String(),
+		"sik", hex.EncodeToString(newSik))
+	// if the hysteresis is reached update the sik for the address
+	v.Tx.Lock()
+	defer v.Tx.Unlock()
+	return v.Tx.DeepSet(address.Bytes(), newSik, StateTreeCfg(TreeSIK))
+}
+
+// DelSIK function removes the registered SIK for the address provided. If it is
+// not registered, it returns an error. If it is, it will encode the hysteresis
+// height and set it as SIK value to invalidate it and prevent it to being
+// updated before that height. The hysteresisHeight must be greater than the
+// curren chain height.
+func (v *State) DelSIK(address common.Address, hysteresisHeight uint32) error {
+	// if the sik does not exists or something fails querying return the error
+	sik, err := v.mainTreeViewer(false).DeepGet(address.Bytes(), StateTreeCfg(TreeSIK))
+	if err != nil {
+		return err
+	}
+	// if the stored sik is already invalidated return an error
+	if !validSIK(sik) {
+		return ErrSIKAlreadyInvalid
+	}
+	// if the provided hysteresis height is lower or equal to the current height
+	// return an error
+	if hysteresisHeight <= v.CurrentHeight() {
+		return ErrInvalidHysteresis
+	}
+	v.Tx.Lock()
+	defer v.Tx.Unlock()
+	return v.Tx.DeepSet(address.Bytes(), encodeHysteresis(hysteresisHeight), StateTreeCfg(TreeSIK))
+}
+
+// encodeHysteresis funtion returns the encoded value of the height hysteresis
```

Comment:
This approach is not correct. It does not avoid double voting.

Instead of an hysteresis height, we should store the the current maximum endBlock for a process.